### ISSUE TITLE
Scoring a translation without decoding?

### ISSUE THREAD

**Issue by MattWiethoff** [OpenNMT/CTranslate2#272]
Hi there. I'd like to be able to score an existing batch of translations (so I already have the outputs and don't need to to do a search). Is there a performant way of doing this?

It'd also be great if I could exploit examples in the batch having shared output prefixes / inputs.

**Comment by guillaumekln** [comment]
Hi,

There is no user-facing API to do that at the moment, but this sounds like something we could add in the future.

Right now it's possible to do that from the code by iteratively calling the decoder and picking the scores of your existing translations. But of course it's a bit of work to get to know the existing code.

By the way, did you already try with the PyTorch or TensorFlow implementation? If yes, what makes you look into CTranslate2?

**Comment by MattWiethoff** [comment]
Thanks for the response, makes sense. That seems like something I take a shot at, but i'm worried there are going to be performance implications to not computing the decoder all at once.

I haven't tried the PyTorch/TF implementations. I've been using CTranslate2 to do low-latency translations on a cpu with minimal dependencies, and was hoping to achieve the same thing for scoring.

**Comment by guillaumekln** [comment]
Iterating over each step is not optimal but it should not make a big difference on CPU unless you are using many cores. It's not too complex to decode at once but this requires a bit of masking work that can get tricky.
